### PR TITLE
Fix #370 - implement proper ER-mode comparison for floating point types

### LINKED ISSUE #370: Seq.sortBy cannot handle sequences of floats containing NaN
I stumbled upon these cases in a one-off fsx - I'm not regularly using F#. This might be entirely my fault.
That said:

Microsoft (R) F# Interactive version 12.0.30815.0
Copyright (c) Microsoft Corporation. All Rights Reserved.

```
> [| 3.2; 1.5; nan; 4.7; nan; 8.5 |] |> Array.sortBy (fun x -> x);;
val it : float [] = [|nan; nan; 1.5; 3.2; 4.7; 8.5|]
```
Good.

```
> [| 3.2; 1.5; nan; 4.7; nan; 8.5 |] |> Seq.sort;;
val it : seq<float> = seq [nan; nan; 1.5; 3.2; ...]
```

Fine.

```
> [| 3.2; 1.5; nan; 4.7; nan; 8.5 |] |> Seq.sortBy (fun x -> x);;
val it : seq<float> = seq [3.2; 1.5; nan; 4.7; ...]
```

What? Why?

I would expect that Array.sortBy and Seq.sortBy behave the same. And I would expect that sortBy identity is the same as sort? What's going on here?

### PR DESCRIPTION
**Pull request by latkin**
Fixes #370. Root cause was that `GenericComparisonFast` did not have an accurate implementation of ER-mode floating point comparison.  Testing showed that nonstructural `compare` had the same issue.

The added `compre` tests here should cover the code well enough, but I'm also planning to add some actual floating-point sorting tests to the core unittests.

### CODE REVIEW

**Comment by forki** [comment]
I merged you changes into #372 and our sorting tests become green ;-)

**Comment by PatrickMcDonald** [comment]
:+1: This is fantastic!  I had no clue how to get around NaNs when doing the SortBy, turns out I was only about 2 levels of abstractions away from enlightenment.

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 1]
```diff
@@ -1284,8 +1290,18 @@ namespace Microsoft.FSharp.Core
...
                  when 'T : uint64 = if (# "clt.un" x y : bool #) then (-1) else (# "cgt.un" x y : int #)
                  when 'T : unativeint = if (# "clt.un" x y : bool #) then (-1) else (# "cgt.un" x y : int #)
-                 when 'T : float  = if (# "clt" x y : bool #) then (-1) else (# "cgt" x y : int #)
-                 when 'T : float32 = if (# "clt" x y : bool #) then (-1) else (# "cgt" x y : int #)
+                 when 'T : float  = if   (# "clt" x y : bool #) then (-1)
+                                    elif (# "cgt" x y : bool #) then (1)
+                                    elif (# "ceq" x y : bool #) then (0)
+                                    elif (# "ceq" x x : bool #) then (1)
+                                    elif (# "ceq" y y : bool #) then (-1)
+                                    else (0)
```
Comment (PatrickMcDonald) [inline src/fsharp/FSharp.Core/prim-types.fs]: Reordering the nan comparisons we can eliminate a branch

    when 'T : float  = if   (# "clt" x y : bool #) then (-1)
                       elif (# "cgt" x y : bool #) then (1)
                       elif (# "ceq" x y : bool #) then (0)
                       elif (# "ceq" y y : bool #) then (-1)
                       else (# "ceq" x x : bool #)

It's similar to what was done here previously, but doesn't avoid any extra IL calls and is arguably less clear

Comment (latkin) [inline src/fsharp/FSharp.Core/prim-types.fs]: Nice! I think you mean `(# "ceq" x x : int #)` on the last line (return the `ceq` result as int value 1 or 0, not bool value)

Comment (PatrickMcDonald) [inline src/fsharp/FSharp.Core/prim-types.fs]: Yes, that's what I meant :blush:

**Comment by dsyme** [comment]
Nice work - great fix, great testing.

**Comment by KevinRansom** [comment]
Looks great

**Pull request closed without merging**